crates/sui-core/src/checkpoints/checkpoint_output.rs

```diff
@@ -106,3 +107,43 @@ impl CertifiedCheckpointOutput for LogCheckpointOutput {
         Ok(())
     }
 }
+
+pub struct SendCheckpointToStateSync {
+    handle: std::sync::Arc<OnceCell<sui_network::state_sync::Handle>>,
+}
+
+impl SendCheckpointToStateSync {
+    pub fn new() -> (
+        Self,
+        std::sync::Arc<OnceCell<sui_network::state_sync::Handle>>,
+    ) {
+        let handle = std::sync::Arc::new(OnceCell::new());
+        (
+            Self {
+                handle: handle.clone(),
+            },
+            handle,
+        )
+    }
+}
+
+#[async_trait]
+impl CertifiedCheckpointOutput for SendCheckpointToStateSync {
+    async fn certified_checkpoint_created(
+        &self,
+        summary: &CertifiedCheckpointSummary,
+    ) -> SuiResult {
+        info!(
+            "Certified checkpoint with sequence {} and digest {}",
+            summary.summary.sequence_number,
+            Hex::encode(summary.summary.digest())
+        );
+        self.handle
+            .get()
+            .ok_or_else(|| SuiError::Unknown("state_sync not initialized yet".into()))?
```

Comment:
I wonder if this error could be avoided if we just put a sender directly into `SendCheckpointToStateSync`?
So right now during startup we do following:
(1) Create `SendCheckpointToStateSync` with empty `handle`
(2) Start `state_sync`, receive `Handle`
(3) Set `Handle` to `SendCheckpointToStateSync`
(4) In the mean time, checkpoint service running between steps (1) and (3) will spit errors to logs and do retries

Instead we could do it this way:
(1) On startup, create pair (checkpoint_sender, checkpoint_receiver)
(2) Pass sender to checkpoint `SendCheckpointToStateSync`
(3) Pass receiver to state_sync initialization

This seem more natural, and this way checkpoint service can start doing progress right away without ever encountering the `state_sync not initialized yet` error from here

Reply:
The only reason that its done this way is that theres the circular dependency where the Checkpoint service and state sync both need access to the checkpoint DB. this could be completely avoided if instead we initialized the DB and passed it to each service. This would let state-sync be constructed first, allowing for the handle to be passed to the checkpoint service without this workaround.

Reply:
But simply constructing channel outside of state sync and passing receiver there would also resolve this circular dependency issue right?

Reply:
It's actually a much larger change than you think it is today. I do have another solution that would also work